src/db/collections/analytics/ledger_size.rs

```diff
@@ -2,116 +2,181 @@
+use super::{Analytic, BlockAnalytics, Error, Measurement, PerMilestone};
 use crate::{
-    db::{
-        collections::{OutputCollection, ProtocolUpdateCollection},
-        MongoDb, MongoDbCollection, MongoDbCollectionExt,
-    },
+    db::MongoDb,
     types::{
-        ledger::RentStructureBytes,
-        stardust::milestone::MilestoneTimestamp,
-        tangle::{MilestoneIndex, RentStructure},
+        ledger::{BlockMetadata, RentStructureBytes},
+        stardust::{
+            block::{payload::TransactionEssence, Block, Output, Payload},
+            milestone::MilestoneTimestamp,
+        },
+        tangle::{MilestoneIndex, ProtocolParameters}, context::TryFromWithContext,
     },
 };
 
+trait StupidRent {
+    fn ledger_stat(&self, protocol_params: &ProtocolParameters) -> LedgerSizeStatistic;
+}
+
+impl StupidRent for Output {
+    fn ledger_stat(&self, protocol_params: &ProtocolParameters) -> LedgerSizeStatistic {
+        let config = iota_types::block::protocol::ProtocolParameters::try_from(protocol_params.clone()).unwrap();
+        let output = iota_types::block::output::Output::try_from_with_context(&config, self.clone()).unwrap();
+        let rent_bytes = RentStructureBytes::compute(&output);
+        LedgerSizeStatistic {
+            total_storage_deposit_value: output.rent_cost(config.rent_structure()),
+            total_key_bytes: rent_bytes.num_key_bytes,
+            total_data_bytes: rent_bytes.num_data_bytes,
+        }
+    }
+}
+
 /// Computes the size of the ledger.
-#[derive(Debug)]
-pub struct LedgerSizeAnalytics;
+#[derive(Debug, Default)]
+pub struct LedgerSizeAnalytics {
+    protocol_params: Option<ProtocolParameters>,
+    stats: LedgerSizeStatistic,
+}
 
-#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
-pub struct LedgerSizeAnalyticsResult {
+#[derive(Copy, Clone, Debug, Default, PartialEq, AddAssign, SubAssign)]
+pub struct LedgerSizeStatistic {
     pub total_key_bytes: u64,
     pub total_data_bytes: u64,
     pub total_storage_deposit_value: u64,
 }
 
+impl BlockAnalytics for LedgerSizeAnalytics {
+    type Measurement = LedgerSizeStatistic;
+
+    fn begin_milestone(&mut self, _index: MilestoneIndex) {}
+
+    fn handle_block(&mut self, block: &Block, _block_metadata: &BlockMetadata, inputs: &Option<Vec<Output>>) {
+        let protocol_params = self.protocol_params.as_ref().unwrap();
+        if let Some(payload) = block.payload.as_ref() {
+            match payload {
+                Payload::Transaction(txn) => {
+                    let TransactionEssence::Regular { outputs, .. } = &txn.essence;
+                    for output in outputs.iter() {
+                        self.stats += output.ledger_stat(protocol_params);
+                    }
+                }
+                _ => (),
+            }
+        }
+        if let Some(inputs) = inputs {
+            for input in inputs {
+                self.stats -= input.ledger_stat(protocol_params);
+            }
+        }
```

Comment:
Reminder: Check if a TX was actually applied

Reply:
thanks, you're absolutely right of course :+1: